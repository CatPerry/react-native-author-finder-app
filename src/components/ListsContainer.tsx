import 'react-native-gesture-handler';
import _ from 'lodash';
import React, { useEffect, useState } from 'react';
import {
  Platform,
  ScrollView,
  StyleSheet,
  TouchableOpacity,
  View,
} from 'react-native';
import { Button, ListItem } from 'react-native-elements';
import { AntDesign } from '@expo/vector-icons';
import PostDetailsModel from '../models/PostDetailsModel';
import PostList from './PostList';

const ListsContainer: React.FunctionComponent = () => {
  const [posts, setPosts] = useState([]);
  const [filteredPosts, setFilteredPosts] = useState([]);
  const [shouldShowBackButton, setshouldShowBackButton] = useState(false);

  const fetchData = async () => {
    const API_PATH =
      Platform.OS === 'android' ? '192.168.1.151:4000' : 'localhost:4000';

    await fetch(`http://${API_PATH}/posts`)
      .then((response) => response.json())
      .then((data) => setPosts(data))
      .catch((error) => console.log(error));
  };

  useEffect(() => {
    fetchData();
  }, []);

  const resetPosts = (): void => {
    setFilteredPosts([]);
    setshouldShowBackButton(false);
  };

  const filterByAuthor = (post: PostDetailsModel): void => {
    let filteredPosts = posts.filter(
      (individual: PostDetailsModel) =>
        individual.author.name === post.author.name
    );
    setFilteredPosts(filteredPosts);
    setshouldShowBackButton(true);
  };

  const showCorrectPosts = (): PostDetailsModel[] => {
    return !_.isEmpty(filteredPosts) ? filteredPosts : posts;
  };

  const renderButton = (): React.ReactNode => {
    return (
      <Button
        title="Back"
        buttonStyle={{
          backgroundColor: '#6169c4',
          justifyContent: 'flex-start',
        }}
        onPress={() => resetPosts()}
        icon={
          <AntDesign
            name="leftcircle"
            size={24}
            color="white"
            style={{ paddingLeft: 5, paddingRight: 9 }}
          />
        }
      />
    );
  };

  return (
    <ScrollView style={{ paddingTop: 25 }}>
      <View>
        {shouldShowBackButton ? renderButton() : null}
        {_.uniqBy(showCorrectPosts(), 'author.name').map(
          (post: PostDetailsModel) => (
            <TouchableOpacity key={post.id}>
              <ListItem
                key={post.id}
                onPress={() => filterByAuthor(post)}
                title={post.author.name}
                bottomDivider
              />
            </TouchableOpacity>
          )
        )}
      </View>
      <PostList posts={showCorrectPosts()} style={styles.postList} />
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  authorList: {
    flexGrow: 2,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  postList: {
    justifyContent: 'flex-start',
  },
  columnView: {
    flex: 1,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    width: '100%',
    marginRight: 50,
  },
});

export default ListsContainer;
