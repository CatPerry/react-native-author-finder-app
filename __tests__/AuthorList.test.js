import React from 'react';
import renderer, { act } from 'react-test-renderer';
import 'isomorphic-fetch';
import MockedNavigator from '../MockedNavigator';
import ListsContainer from '../src/components/ListsContainer';

test('ListsContainer fetches data correctly', async () => {
  const promise = Promise.resolve();
  const handleFetchData = jest.fn(() => promise);
  const wrapper = renderer
    .create(
      <MockedNavigator
        component={ListsContainer}
        handleFetchData={handleFetchData}
      />
    )
    .toJSON();

  expect(wrapper).toMatchSnapshot();
  await act(() => promise);
});
