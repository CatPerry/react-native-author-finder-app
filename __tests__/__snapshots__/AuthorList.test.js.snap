// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsContainer fetches data correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          false,
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(224, 224, 224)",
                "flex": 1,
                "shadowColor": "rgb(224, 224, 224)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                MockedMainScreen
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
                "opacity": 1,
              }
            }
          >
            <View
              collapsable={false}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <RCTScrollView
                      style={
                        Object {
                          "paddingTop": 25,
                        }
                      }
                    >
                      <View>
                        <View />
                        <RCTSafeAreaView
                          emulateUnlessSupported={true}
                          style={
                            Object {
                              "justifyContent": "flex-start",
                              "width": "100%",
                            }
                          }
                        >
                          <RCTScrollView>
                            <View />
                          </RCTScrollView>
                        </RCTSafeAreaView>
                      </View>
                    </RCTScrollView>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
